### PR TITLE
[avmfritz] Added new channel 'device_locked' and other improvements

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added `UnDefType.UNDEF` state for channel `next_temp` if AHA sends undefined value
- Added new channel `device_locked`
- Added `model id` and `vendor` properties to discovered things
- Added `updateProperty()` method for `PROPERTY_FIRMWARE_VERSION`
- Removed 'com.google.common.collect' dependency
- Removed helper classes for discovery and polling jobs
- Fixed use of representation properties

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java [hunk 1]
```diff
@@ -161,6 +157,7 @@ private void updateThingFromDevice(Thing thing, DeviceModel device) {
         }
         if (device.getPresent() == 1) {
             thing.setStatusInfo(new ThingStatusInfo(ThingStatus.ONLINE, ThingStatusDetail.NONE, null));
+            updateThingProperty(thing, Thing.PROPERTY_FIRMWARE_VERSION, device.getFirmwareVersion());
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java]: I would love static imports for that :-)

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java [hunk 2]
```diff
@@ -201,8 +204,12 @@ private void updateThingFromDevice(Thing thing, DeviceModel device) {
                         calendar.setTime(new Date(device.getHkr().getNextchange().getEndperiod() * 1000L));
                         updateThingChannelState(thing, CHANNEL_NEXTCHANGE, new DateTimeType(calendar));
                     }
-                    updateThingChannelState(thing, CHANNEL_NEXTTEMP,
-                            new DecimalType(HeatingModel.toCelsius(device.getHkr().getNextchange().getTchange())));
+                    if (HeatingModel.TEMP_FRITZ_UNDEFINED.equals(device.getHkr().getNextchange().getTchange())) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java]: Should you maybe put this logic in a method it seems like something you could need more often.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java [hunk 3]
```diff
@@ -234,6 +241,19 @@ private void updateThingChannelState(Thing thing, String channelId, State state)
...
 
     /**
+     * Updates thing properties.
+     * 
+     * @param thing Thing which property to be updated.
+     * @param propertyId ID of the property to be updated.
+     * @param propertyValue Value to be set.
+     */
+    private void updateThingProperty(Thing thing, @NonNull String propertyId, String propertyValue) {
+        if (thing.getProperties().containsKey(propertyId)) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java]: Why this check?

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java [hunk 4]
```diff
@@ -353,7 +366,7 @@ private void updateThingFromDevice(Thing thing, DeviceModel device) {
     /**
      * Updates thing channels.
      *
-     * @param thing Thing to be updated.
+     * @param thing Thing which channel to be updated.
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java]: Maybe 'Thing which channels should be updated'?

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java [hunk 5]
```diff
@@ -367,6 +380,19 @@ private void updateThingChannelState(Thing thing, String channelId, State state)
...
 
     /**
+     * Updates thing properties.
+     * 
+     * @param thing Thing which property to be updated.
+     * @param propertyId ID of the property to be updated.
+     * @param propertyValue Value to be set.
+     */
+    private void updateThingProperty(Thing thing, @NonNull String propertyId, String propertyValue) {
+        if (thing.getProperties().containsKey(propertyId)) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java]: See earlier remark, if this check is really needed is there some re-use possible?

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzahaWebInterface.java [hunk 6]
```diff
@@ -91,7 +91,7 @@
     public String authenticate() {
         if (this.config.getPassword() == null) {
             this.fbHandler.setStatusInfo(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "please configure password first");
+                    "no password set");
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzahaWebInterface.java]: What was wrong with the old text?

**Comment by cweitkamp** [comment]
> what is wrong with de_DE properties according to GitHub they are binary.

@martinvw I saw it in the same moment you posted your review. I don't know what really happened, but I somehow have the feeling that the file got broken when resolving merge conflicts using GitHub.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzahaWebInterface.java (hunk 6 above)
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzahaWebInterface.java]: To be honest: I really can't remember why I changed it. Will revert the change.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java (hunk 5 above)
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/DeviceHandler.java]: I don't think this check is necessary. My intention was not to update a property which doesn't exist. But that is nonsense.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java [hunk 7]
```diff
@@ -113,9 +110,10 @@ public void onDeviceAddedInternal(DeviceModel device) {
...
-            properties.put(Thing.PROPERTY_VENDOR, device.getManufacturer());
-            properties.put(Thing.PROPERTY_SERIAL_NUMBER, device.getIdentifier());
-            properties.put(Thing.PROPERTY_FIRMWARE_VERSION, device.getFirmwareVersion());
+            properties.put(PROPERTY_VENDOR, device.getManufacturer());
+            properties.put(PROPERTY_MODEL_ID, device.getDeviceId());
+            properties.put(PROPERTY_SERIAL_NUMBER, device.getIdentifier());
+            properties.put(PROPERTY_FIRMWARE_VERSION, device.getFirmwareVersion());
 
             final DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withProperties(properties)
                     .withRepresentationProperty(device.getIdentifier()).withBridge(bridgeUID)
```
Comment (cweitkamp) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: @martinvw I have a question:  When using `withRepresentationProperty()` method, what do I have to pass as parameter?  The `name` of the property or the `value` for the representational property?

Comment (martinvw) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: Good catch, see https://github.com/eclipse/smarthome/blob/master/bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/DiscoveryResult.java#L77

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: Thanks. I fixed it.

**Comment by martinvw** [comment]
@sjka can you check for the correct usage of representation-property in the discovery and xml?

**Comment by sjka** [comment]
In the thing types XMLs you define _which_ property is "the one" (i.e. "ain") and this property is filled with the identifying _value_ in the discovery service. Looks good to me!

**Comment by martinvw** [comment]
Thanks to both of you!

**Pull request merged**